### PR TITLE
Add option to run e2e tests locally

### PR DESCRIPTION
**Pull request by haonliu**
#### Description of changes

Add local e2e test script.

Provide the option to run functional test locally before check in. 

- if we make some changes to az function, instead of deploy the whole service and run the dev pipeline to check the functional test results, we can just publish the function and then run the e2e tests to verify the changes. 
- when we only changed the client side code, we don't need to deploy anything, just run the e2e tests locally.
- when we add more test cases into e2e tests, we can verify them easily.
- make it easier for debugging purpose.

#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Added relevant unit test for your changes. (`yarn test`)
- [ ] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [ ] Ran precheckin (`yarn precheckin`)

### CODE REVIEW

File: packages/functional-tests/src/functional-tests.local.spec.ts [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+import { FunctionalTestGroup } from './test-groups/functional-test-group';
+
+// tslint:disable: mocha-no-side-effect-code no-any no-unsafe-any mocha-unneeded-done strict-boolean-expressions
+
+describe('functional tests', () => {
+    // need to provide by user. DO NOT CHECK IN
+    const clientId = process.env.SP_CLIENT_ID || '';
+    const clientSecret = process.env.SP_PASSWORD || '';
+    const tenantId = process.env.SP_TENANT || '';
+    const apimName = process.env.APIM_Name || '';
```
Comment (lamaks) [inline packages/functional-tests/src/functional-tests.local.spec.ts]: APIM_SERVICE_NAME

File: packages/functional-tests/src/functional-tests.local.spec.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    let a11yServiceClient: A11yServiceClient;
+    let testRunner: TestRunner;
+    let guidGenerator: GuidGenerator;
+    let testContextData: TestContextData;
+    let cosmosClientWrapper: CosmosClientWrapper;
+    let cosmosContainerClient: CosmosContainerClient;
+    let onDemandPageScanRunResultProvider: OnDemandPageScanRunResultProvider;
+
+    beforeAll(async () => {
+        if (isServiceCredProvided()) {
```
Comment (lamaks) [inline packages/functional-tests/src/functional-tests.local.spec.ts]: if not provided do we want to indicate that?

File: packages/functional-tests/src/functional-tests.local.spec.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+            testRunner = new TestRunner(logger);
+            guidGenerator = new GuidGenerator();
+            cosmosClientWrapper = new CosmosClientWrapper(async () => {
+                return new CosmosClient({ endpoint: cosmosUrl, auth: { masterKey: cosmosKey } });
+            }, logger);
+            cosmosContainerClient = new CosmosContainerClient(cosmosClientWrapper, dbName, collectionName, logger);
+            onDemandPageScanRunResultProvider = new OnDemandPageScanRunResultProvider(
+                cosmosContainerClient,
+                new PartitionKeyFactory(new HashGenerator(), guidGenerator),
+            );
```
Comment (lamaks) [inline packages/functional-tests/src/functional-tests.local.spec.ts]: It seems  most of this code already implemented when we setup DI container. Can we reuse DI container implementation?

File: packages/functional-tests/src/functional-tests.local.spec.ts [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    const clientSecret = process.env.SP_PASSWORD || '';
+    const tenantId = process.env.SP_TENANT || '';
+    const apimName = process.env.APIM_Name || '';
+    const cosmosKey = process.env.COSMOS_DB_KEY || '';
+    const cosmosUrl = process.env.COSMOS_DB_URL || '';
+    // static values
+    const dbName = 'onDemandScanner';
+    const collectionName = 'scanRuns';
+    const releaseIdStub = 'release-id';
+    const runIdStub = 'run-id';
```
Comment (lamaks) [inline packages/functional-tests/src/functional-tests.local.spec.ts]: Better to have real values here since it goes to app insight and will have to debug

Comment (haonliu) [reply packages/functional-tests/src/functional-tests.local.spec.ts]: updated.

File: packages/functional-tests/src/functional-tests.local.spec.ts (hunk 3 above)
Comment (haonliu) [reply packages/functional-tests/src/functional-tests.local.spec.ts]: Updated.

Comment (haonliu) [reply packages/functional-tests/src/functional-tests.local.spec.ts]: Reused some DI implementation that is available. Need to refactor the 'getProcessLifeCycleContainer' in web-worker before we can fully reuse that. will do that in a separate PR.

File: packages/functional-tests/src/functional-tests.local.spec.ts [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        const scanReportsTests = getTests('ScanReports');
+        await testRunner.run(scanReportsTests, TestEnvironment.all, releaseIdStub, runIdStub);
+        done();
+    });
+
+    testIf('Finalizer', isServiceCredProvided, async (done: jest.DoneCallback) => {
+        const finalizerTests = getTests('Finalizer');
+        await testRunner.run(finalizerTests, TestEnvironment.all, releaseIdStub, runIdStub);
+        done();
+    });
```
Comment (lamaks) [inline packages/functional-tests/src/functional-tests.local.spec.ts]: It looks very similar implementation. Can we do some generic function that get list of tests groups to run in a sequence?

File: packages/functional-tests/src/functional-tests.local.spec.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    testIf('ScanStatus', isServiceCredProvided, async (done: jest.DoneCallback) => {
+        const scanStatusTests = getTests('ScanStatus');
+        testContextData.scanId = (await a11yServiceClient.postScanUrl(testContextData.scanUrl)).body[0].scanId;
+        await testRunner.run(scanStatusTests, TestEnvironment.all, releaseIdStub, runIdStub);
+        done();
+    });
+
+    testIf('ScanExecution', isServiceCredProvided, async (done: jest.DoneCallback) => {
+        const scanRunResultResponse = await waitForScanRequestCompletion();
+        expect(scanRunResultResponse.reports.length).toEqual(2);
```
Comment (lamaks) [inline packages/functional-tests/src/functional-tests.local.spec.ts]: Should this function just run e2e tests w/o any validation? If we need validation it should be part of an e2e test instead

File: packages/functional-tests/src/functional-tests.local.spec.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+import { TestEnvironment } from './common-types';
+import { functionalTestGroupTypes, TestGroupName } from './functional-test-group-types';
+import { TestRunner } from './runner/test-runner';
+import { TestContextData } from './test-group-data';
+import { FunctionalTestGroup } from './test-groups/functional-test-group';
+
+// tslint:disable: mocha-no-side-effect-code no-any no-unsafe-any mocha-unneeded-done strict-boolean-expressions
+
+describe('functional tests', () => {
```
Comment (lamaks) [inline packages/functional-tests/src/functional-tests.local.spec.ts]: In general would be better to have this implemented as a generic e2e tests group runner similar to what we have in health orchestration worker

Comment (lamaks) [reply packages/functional-tests/src/functional-tests.local.spec.ts]: Reuse the health REST API result can be part of this, so we can unify implementation and do not fork e2e runners

File: packages/functional-tests/src/functional-tests.local.spec.ts (hunk 5 above)
Comment (haonliu) [reply packages/functional-tests/src/functional-tests.local.spec.ts]: yep. updated.

File: packages/functional-tests/src/functional-tests.local.spec.ts (hunk 6 above)
Comment (haonliu) [reply packages/functional-tests/src/functional-tests.local.spec.ts]: removed.

File: packages/functional-tests/src/functional-tests.local.spec.ts (hunk 7 above)
Comment (haonliu) [reply packages/functional-tests/src/functional-tests.local.spec.ts]: Agreed. that could be in a future PR?

**Pull request merged**